lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,964 @@
+//===- ExtractInstances.cpp - Move instances up the hierarchy ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Moves annotated instances upwards in the module hierarchy. Corresponds to the
+// `ExtractBlackBoxes`, `ExtractClockGates`, and `ExtractSeqMems` passes in the
+// Scala FIRRTL implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/NLATable.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWDialect.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/Path.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Support/FileUtilities.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+
+#define DEBUG_TYPE "firrtl-extract-instances"
+
+using namespace circt;
+using namespace firrtl;
+using hw::InnerRefAttr;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// All information necessary to move instances about.
+struct ExtractionInfo {
+  /// A filename into which the performed hierarchy modifications are emitted.
+  StringRef traceFilename;
+  /// A prefix to attach to the wiring generated by the extraction.
+  StringRef prefix;
+  /// Optional name of the wrapper module that will hold the moved instance.
+  StringRef wrapperModule;
+  /// Whether the extraction should stop at the root of the DUT instead of going
+  /// past that and extracting into the test harness.
+  bool stopAtDUT;
+};
+
+struct ExtractInstancesPass
+    : public ExtractInstancesBase<ExtractInstancesPass> {
+  void runOnOperation() override;
+  void collectAnnos();
+  void collectAnno(InstanceOp inst, Annotation anno);
+  void extractInstances();
+  void groupInstances();
+  void createTraceFiles();
+
+private:
```

Comment:
there are back-to-back `private:` labels, although i'm not sure you really need to bother marking anything private. Possibly only `runOnOperation()` should be public?

Reply:
Yeah good point, it's internal anyway. Removed the `private`.